### PR TITLE
refactor: Deref `MainTrace` type to `ColMatrix`

### PR DESCRIPTION
**Pull request by iammadab**
## Describe your changes
- Implement Deref for MainTrace to &ColMatrix
- Remove lifetime annotation from Maintrace
   - new method now takes in ColMatrix as opposed to &ColMatrix
- Replace &ColMatrix with &MainTrace  

relevant issue: #1201

### CODE REVIEW

File: air/src/trace/main_trace.rs [hunk 1]
```diff
@@ -17,6 +17,7 @@ use super::{
     STACK_TRACE_OFFSET,
 };
 use core::ops::Range;
+use std::ops::Deref;
```
Comment (plafer) [inline air/src/trace/main_trace.rs:20]: This should be `core::ops::Deref` (added to the previous line)

File: processor/src/trace/mod.rs [hunk 2]
```diff
@@ -86,7 +87,7 @@ impl ExecutionTrace {
         Self {
             meta: Vec::new(),
             layout: TraceLayout::new(TRACE_WIDTH, [AUX_TRACE_WIDTH], [AUX_TRACE_RAND_ELEMENTS]),
-            main_trace: ColMatrix::new(main_trace),
+            main_trace: MainTrace::new(ColMatrix::new(main_trace)),
```
Comment (plafer) [inline processor/src/trace/mod.rs:90]: `finalize_trace()` should return a `MainTrace` instead of `Vec<Vec<Felt>>` in the first field of the tuple

**Review by plafer — COMMENTED** [review]
Looks great, thank you! Let a few minor nits.

Let's fix the CI too. You can fix by running locally:

```
// fix any clippy errors
$ cargo +nightly clippy --all -- -D clippy::all -D warning
// format once all fixes are done
$ cargo fmt
```

File: processor/src/decoder/tests.rs [hunk 3]
```diff
@@ -1202,7 +1201,7 @@ fn build_trace(stack_inputs: &[u64], program: &CodeBlock) -> (DecoderTrace, usiz
     process.execute_code_block(program, &CodeBlockTable::default()).unwrap();
 
     let (trace, _, _) = ExecutionTrace::test_finalize_trace(process);
-    let trace_len = get_trace_len(&trace) - ExecutionTrace::NUM_RAND_ROWS;
+    let trace_len = trace.num_rows() - ExecutionTrace::NUM_RAND_ROWS;
 
     (
         trace[DECODER_TRACE_RANGE]
```
Comment (iammadab) [inline processor/src/decoder/tests.rs:1207]: @plafer Thinking about how to handle MainTrace indexed at a range. 
1. Thought of implementing get_column_range(range: Range<usize>) -> &[..] in the MainTrace impl
2. Another option is to implement dedicated methods for retrieving those columns
    - DECODER_TRACE_RANGE
    - SYS_TRACE_RANGE
    - CHIPLETS_RANGE

What do you think? also, is there a better way?

Comment (iammadab) [inline processor/src/decoder/tests.rs:1207]: @plafer Thinking about how to handle MainTrace indexed at a range. 
1. Thought of implementing get_column_range(range: Range<usize>) -> &[..] in the MainTrace impl
2. Another option is to implement dedicated methods for retrieving those columns
    - DECODER_TRACE_RANGE
    - SYS_TRACE_RANGE
    - CHIPLETS_RANGE

What do you think? also, is there a better way?

Comment (iammadab) [reply processor/src/decoder/tests.rs]: decided to go with the first approach

Comment (plafer) [reply processor/src/decoder/tests.rs]: Yes I prefer the first approach too

File: processor/src/chiplets/tests.rs [hunk 4]
```diff
@@ -117,11 +117,11 @@ fn build_trace(
...
 
     let (trace, _, _) = ExecutionTrace::test_finalize_trace(process);
-    let trace_len = get_trace_len(&trace) - ExecutionTrace::NUM_RAND_ROWS;
+    let trace_len = trace.num_rows() - ExecutionTrace::NUM_RAND_ROWS;
 
     (
-        trace[CHIPLETS_RANGE]
-            .to_vec()
+        trace
+            .get_column_range(CHIPLETS_RANGE)
```
Comment (plafer) [inline processor/src/chiplets/tests.rs:124]: Here we end up cloning some columns of `trace`, and then dropping `trace`, which isn't great. But the previous implementation did that, so we don't worsen the performance, and we can leave it for another PR to improve on it. And the compiler might even optimize that `clone()` away anyways.

**Review by plafer — APPROVED** [review]
LGTM!

@bobbinth want to take a look before we merge?

**Comment by plafer** [comment]
@iammadab oops can you fix the CI? Basically an explicit import of `Vec` is needed for `no_std` (look at how other files do it)

**Comment by iammadab** [comment]
@plafer pushed no-std fix, we can run CI again.

File: processor/src/chiplets/tests.rs (hunk 4 above)
Comment (bobbinth) [reply processor/src/chiplets/tests.rs:124]: Yes - not critical to improve, especially since this is just for testing.

File: air/src/trace/main_trace.rs [hunk 5]
```diff
@@ -43,6 +54,13 @@ impl<'a> MainTrace<'a> {
         self.columns.num_rows()
     }
 
+    pub fn get_column_range(&self, range: Range<usize>) -> Vec<Vec<Felt>> {
+        range.fold(vec![], |mut acc, col_idx| {
+            acc.push(self.get_column(col_idx).to_vec());
+            acc
+        })
+    }
```
Comment (bobbinth) [inline air/src/trace/main_trace.rs:62]: This could be a very expensive method as it clones trace columns. However, as far as I can tell, this is used only for testing purposes - so, this should be fine. However, ideally, we'd put this method behind a feature.

Currently, we run tests under  the `internals` feature (which we should probably rename to `testing` - but that's for another PR). Not sure how straight-forward would it be to add this feature here as well. If it is not straight-forward, we can do it in a future PR and just leave a `TODO` here for now to make sure it is clear that this method is for testing purposes only.

**Review by bobbinth — COMMENTED** [review]
Looks great! Thank you for working on this! I left one comment inline - but it could be as simple as adding a comment to the relevant method.

File: air/src/trace/main_trace.rs (hunk 5 above)
Comment (iammadab) [reply air/src/trace/main_trace.rs:62]: put the get_column_range function behind the internals feature.

**Comment by plafer** [comment]
The clippy error is due to the latest `nightly` clippy (all branches have it). I think we're good to merge this with the error, and fix it elsewhere (and only once).

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you again!
